### ISSUE TITLE
Interaction between opam and dune broken w/ (opam_file_location inside_opam_directory)

### ISSUE THREAD

**Issue by jonahbeckford** [ocaml/dune#9727]
## Expected Behavior

I realize that `(opam_file_location inside_opam_directory)` is not being advertised, but I think the `opam_file_location` feature will cause problems if people start using it. Consider this a tracking issue. This is marginally related to open issue #7534 as well.

With a `dune-project` file:

```scheme
(lang dune 3.12)

(name DkSDKMetatype)
(version 1.0.0)

(generate_opam_files true)
(opam_file_location inside_opam_directory)

(package
  (name DkSDKMetatype_Common)
  (depends
    (ocaml (>= 4.14.0))
    (semver2 (>= 1.2.0))))
```

and a basic `dune` file:

```scheme
(documentation
 (package DkSDKMetatype_Common))

(include_subdirs unqualified)

(library
 (name DkSDKMetatype_Common)
 (public_name DkSDKMetatype_Common)
 (libraries semver2)
 (instrumentation (backend bisect_ppx)))
```

when I run:

```sh
opam exec -- dune build '@opam/all' # generates opam/ folder correctly
opam install ./opam/DkSDKMetatype_Common.opam --verbose --debug-level 2
```

I expect the `OPAMSWITCH/lib/DkSDKMetatype_Common/` folder to be populated with compiled artifacts like `DkSDKMetatype_Common.cma`.

## Actual Behavior

Instead, I get only three files:

```
Y:\SOURCE\DKSDK-TYPE\_OPAM\LIB\DKSDKMETATYPE_COMMON
    dune-package
    META
    opam
```

What is happening is obvious from the [opamlog.txt `opam install` logs](https://github.com/ocaml/dune/files/13929929/opamlog.txt):

> `rsync ... /y/source/dksdk-type/opam/ /y/source/dksdk-type/_opam/.opam-switch/sources/DkSDKMetatype_Common`

Instead of copying the real source folder, `opam` is copying the `opam/` subfolder. And then launching a `dune build` from inside that (incorrect) source folder.

## Mitigation

Either:

* `opam install .` works. But that installs all of the packages in the source tree which is sometimes undesirable.
* Pinning the packages to the source directory (ie. `git+file://$PWD/.git`) and then installing whichever packages you want.

## Specifications

- Version of `dune` (output of `dune --version`): 3.12.1
- Version of `ocaml` (output of `ocamlc --version`)
- opam version: `2.2.0~alpha0~20221228`
- Operating system (distribution and version): Windows 11 Pro

**Comment by rgrinberg** [comment]
Yes, we're aware of this limitation and this is the reason why we're not using it in dune yet. AFIAK, this is an issue in opam that needs to be addressed first. cc @rjbout to confirm
